### ISSUE TITLE
"base" equals "this" in overriden method

### ISSUE THREAD

**Issue by workerclass** [pamidur/aspect-injector#42]
I have two classes: Base and Child. There is a method in Base that's overriden in Child and that has an Around aspect. Overriden method calls base implementation, but it calls himself instead resulting in StackOverflow exception. 
`base.GetType() == this.GetType()` returns `true`

**Comment by YuriyIvon** [comment]
Thanks for the report, potentially there might be other side-effects with around aspects and virtual functions. We will analyze all possible issues to provide a reliable solution.

**Comment by YuriyIvon** [comment]
I was not able to reproduce StackOverflow issue using a simple example:
```
    class TraceAspect
    {
        [Advice(InjectionPoints.Around, InjectionTargets.Method)]
        public object Trace(
            [AdviceArgument(AdviceArgumentSource.Type)] Type type,
            [AdviceArgument(AdviceArgumentSource.Name)] string methodName,
            [AdviceArgument(AdviceArgumentSource.Target)] Func<object[], object> target,
            [AdviceArgument(AdviceArgumentSource.Arguments)] object[] arguments)
        {
            Console.WriteLine("Before {0}.{1}", type, methodName);
            var result = target(arguments);
            Console.WriteLine("After {0}.{1}", type, methodName);

            return result;
        }
    }

    class Parent
    {
        public virtual void Print()
        {
            Console.WriteLine("In Parent.Print");
        }
    }

    [Aspect(typeof(TraceAspect))]
    class Child : Parent
    {
        public override void Print()
        {
            base.Print();

            Console.WriteLine("In Child.Print");
        }
    }
```

I'm a getting correct result when executing this code, though there are some redundant "override" attributes on generated methods which doesn't have corresponding virtual functions in the parent class. That should be fixed, but currently I have no idea how that can lead to the stack overflow issue.

As for base.GetType() == this.GetType() - this condition will always return true, regardless of AspectInjector, because GetType() returns the exact runtime type of the current instance. Since both "base" and "this" belong to the same instance, the call will return the same type in both cases.

Can you provide more details about .NET framework version and which runtime you are using (Microsoft .NET, Mono)?

**Comment by workerclass** [comment]
You need to add [Aspect(typeof(TraceAspect))] to Parent class too to get the exception. (Microsoft .NET 4.5.2)

**Comment by YuriyIvon** [comment]
I have found the root cause, will publish a fix by tomorrow

**Comment by pamidur** [comment]
Fixed in 1.0.0-beta4
